Hotfix/16974 can see description diff without authorization

floriank opened the pull request:
This should fix a problem where any user could view the description diff of a Work Package.

This was reported in https://community.openproject.org/work_packages/16974

Commit:
Fix error where anyone could read the diff of a WP

With this fix it is no longer possible to view a description diff
without the necessary permissions.

The error was reported by Jan Sandbrink in
https://community.openproject.org/work_packages/16974.

Commit:
Fix the coding style via rubocop

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://floriank-hotfix-16974-can-see-description-diff-without-authoriz.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://floriank-hotfix-16974-can-see-description-diff-without-authoriz.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

myabc commented on spec/controllers/journals_controller_spec.rb:
```diff
@@ -28,49 +28,70 @@
+        expect(response).to be_success
+      end
+
+      it 'should presetn the diff correctly' do
```
@floriank typo

myabc replied:
(not yours though :wink:)

ulferts commented:
@floriank this fixes the issue but raises a new one. The :diff action on the journals controller is currently only accessible for users having the :view_work_packages permission.

The diff method on the other hand aims to be generic, in the sense of it not being limited to returning the diff of work package description (although the visible_field? method already limits this). With the permission being what it is, this now fully becomes limited to the work package diff use case.

I am willing to live with the limitation if we make it explicit. Therefore I propose to return a 404 if the journable/journable_type isn't a work package.

Commit:
fix controller response

As the diff action claims to be more generic, react with a 404 instead
of a 403 whenever the journable type is not a WorkPackage.

This makes the the use of the diff action less generic, as a user will
still need the view_work_pacakges permission to access the diff action.
